modules/tensor_mechanics/src/materials/ScalarDamageBase.C

```diff
@@ -30,16 +30,16 @@ ScalarDamageBase::validParams()
       0.1,
       "maximum_damage_increment>0 & maximum_damage_increment<1",
       "maximum damage increment allowed for simulations with adaptive time step");
-  params.addParam<std::string>("damage_index_name",
+  params.addParam<MaterialPropertyName>("damage_index_name",
                                "damage_index",
                                "name of the material property where the damage index is stored");
   return params;
 }
 
 ScalarDamageBase::ScalarDamageBase(const InputParameters & parameters)
   : DamageBase(parameters),
-    _damage_index_name(getParam<std::string>("damage_index_name")),
-    _damage_index(declareProperty<Real>(_base_name + _damage_index_name)),
+    _damage_index_name(getParam<MaterialPropertyName>("damage_index_name")),
+    _damage_index(declarePropertyByName<Real>(_base_name + _damage_index_name)),
```

Comment:
This change would need to be made in a million other classes as well.

Reply:
...? This is the only one that popped up in modules and apps?

Reply:
It's typically fine everywhere else. the `ByName` is only required when there is a weird input params conflict.

Reply:
`declarePropertyByName` bypasses the automatic name resolution from input parameters. Similar to `getMaterialProperty` and `getMaterialPropertyByName`...

Reply:
Yeah, I'm aware of how that works. I also appreciate this API homogenization. I also think that if the developer knows they are passing in a name they should _always_ use the ByName method, and I'd actually prefer if there wasn't that voodoo magic that behind the scenes allows the argument to mean two entirely different things.

It's just confusing.

Reply:
The "homogenization" is what I'm going for. I don't mind either way, but it's really confusing when `get` and `declare` do different things.

So you just griping or proposing a different way :)

Reply:
Unless you're proposing to replace all existing `declareProperty<T>("some_name")` with `declarePropertyByName<T>("some_name")`, and `declareProperty<T>(getParam<MaterialPropertyName>("some_param"))` with `declareProperty<T>("some_param")`....

Reply:
That would be my preference. But that's a bit beyond the scope of this PR. I'd love if we could deprecate the use of `declareProperty` and `getMaterialProperty` where the argument is _not_ a valid input parameter name.

Reply:
Well, this PR has certainly come a loooooong way from it's original doomsday beginnings....